proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+  (wp: crunch_wps simp: crunch_simps)
+
+definition replyPush' where
+  "replyPush' callerPtr calleePtr replyPtr canDonate scPtrOptDonated scPtrOptCallee \<equiv>
+   do
+       canDonate <- return (if scPtrOptCallee \<noteq> Nothing then False else canDonate);
+       reply <- getReply replyPtr;
+       haskell_assert (replyPrev reply = Nothing) [];
+       haskell_assert (replyNext reply = Nothing) [];
+       tsCaller <- getThreadState callerPtr;
+       haskell_assert (replyObject tsCaller = Nothing) [];
+       tsCallee <- getThreadState calleePtr;
+       haskell_assert (replyObject tsCallee = Nothing) [];
+       setReplyTCB (Just callerPtr) replyPtr;
+       setThreadState (Structures_H.thread_state.BlockedOnReply (Just replyPtr)) callerPtr;
+       when (scPtrOptDonated \<noteq> Nothing \<and> canDonate) $
+       do haskell_assert (scPtrOptCallee = Nothing) [];
+          scDonated <- getSchedContext (fromJust scPtrOptDonated);
+          oldReplyPtrOpt <- return $ scReply scDonated;
+          when (oldReplyPtrOpt \<noteq> Nothing) $
+          do oldReplyPtr <- return $ fromJust oldReplyPtrOpt;
+             oldReply <- getReply oldReplyPtr;
+             haskell_assert (replyNext oldReply = Just (Head $ fromJust scPtrOptDonated)) []
+          od;
+          reply' <- getReply replyPtr;
+          setReply replyPtr
+           (replyNext_update (\<lambda>_. Just (Head $ fromJust scPtrOptDonated))
+             (replyPrev_update (\<lambda>_. oldReplyPtrOpt) reply'));
+          when (oldReplyPtrOpt \<noteq> Nothing) $
+          do oldReplyPtr <- return $ fromJust oldReplyPtrOpt;
+             oldReply <- getReply oldReplyPtr;
+             setReply oldReplyPtr (replyNext_update (\<lambda>_. Just (Next replyPtr)) oldReply)
+          od;
+          scDonated <- getSchedContext (fromJust scPtrOptDonated);
+          setSchedContext (fromJust scPtrOptDonated) (scReply_update (\<lambda>_. Just replyPtr) scDonated)
+       od
+    od"
+
+lemma monadic_rewrite_replyPush:
+  "monadic_rewrite False True \<top>
+   (do stateAssert sym_refs_asrt [];
+       scPtrOptDonated <- threadGet tcbSchedContext callerPtr;
+       tptrOpt <- getReplyTCB replyPtr;
+       haskell_assert (tptrOpt = Nothing) [];
+       scPtrOptCallee <- threadGet tcbSchedContext calleePtr;
+       replyPush' callerPtr calleePtr replyPtr canDonate scPtrOptDonated scPtrOptCallee;
+       when (scPtrOptDonated \<noteq> Nothing \<and> (if scPtrOptCallee \<noteq> Nothing then False else canDonate)) $
+           schedContextDonate (fromJust scPtrOptDonated) calleePtr od)
+   (replyPush callerPtr calleePtr replyPtr canDonate)"
+  unfolding replyPush_def replyPush'_def
+  apply (clarsimp simp only: K_bind_def bind_assoc)
+  apply (rule monadic_rewrite_gen_asm[simplified pred_conj_def]
+              monadic_rewrite_bind_tail[OF _ return_sp]
+              monadic_rewrite_bind_tail[OF _ hoare_post_taut])+
+  apply (case_tac "(scPtrOptDonated \<noteq> Nothing \<and> x)")
```

Comment:
What's x? Could you do a rename_tac?

Reply:
> What's x? Could you do a rename_tac?

It's a bit hard to rename this one; I don't know what name I could give to `(if scPtrOptCallee ≠ Nothing then False else canDonate)`, and there are 9 other variables after `x` that I'd have to rename too unless there's a way to target one variable in particular.

Reply:
Ah. In that case, don't worry about it